exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/BloomFilter.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.work.filter;
+
+import io.netty.buffer.DrillBuf;
+import java.util.Arrays;
+
+
+/**
+ * According to Putze et al.'s "Cache-, Hash- and Space-Efficient BloomFilter
+ * Filters", the main theory is to construct tiny bucket bloom filters which
+ * benefit to the cpu cache and SIMD opcode.
+ */
+
+public class BloomFilter {
+  // Bytes in a bucket.
+  private static final int BYTES_PER_BUCKET = 32;
+  // Minimum bloom filter data size.
+  private static final int MINIMUM_BLOOM_SIZE = 256;
+
+  private static final int DEFAULT_MAXIMUM_BLOOM_FILTER_SIZE = 16 * 1024 * 1024;
+
+  private DrillBuf byteBuf;
+
+  private int numBytes;
+
+  private int mask[] = new int[8];
+
+  private byte[] tempBucket = new byte[32];
+
+
+  public BloomFilter(DrillBuf byteBuf) {
+    this.byteBuf = byteBuf;
+    this.numBytes = byteBuf.capacity();
+    this.byteBuf.writerIndex(numBytes);
+  }
+
+
+  public static int adjustByteSize(int numBytes) {
+    if (numBytes < MINIMUM_BLOOM_SIZE) {
+      numBytes = MINIMUM_BLOOM_SIZE;
+    }
+
+    if (numBytes > DEFAULT_MAXIMUM_BLOOM_FILTER_SIZE) {
+      numBytes = DEFAULT_MAXIMUM_BLOOM_FILTER_SIZE;
+    }
+
+    // 32 bytes alignment, one bucket.
+    numBytes = (numBytes + 0x1F) & (~0x1F);
+    return numBytes;
+  }
+
+  private void setMask(int key) {
+    final int SALT[] = {0x47b6137b, 0x44974d91, 0x8824ad5b, 0xa2b7289d, 0x705495c7, 0x2df1424b, 0x9efc4947, 0x5c6bfb31};
```

Comment:
those 8 odd numbers was picked from Impala's implementation(see bloom-filter.h) . They act as salt values to participate in calculating the mask. I will add some comments to this method to make it clear.

Reply:
Although I understand these are open source software but in general it is better to avoid referencing another project's implementation.  In this case, it is probably ok since picking these values is similar to picking prime numbers for hash computations and typically they are common across implementations.